### PR TITLE
chore(build): use a Filter plugin to write the dest folder.

### PR DESCRIPTION
**Pull request by alexeagle**
This lets broccoli keep the dest folder up-to-date in 'watch' mode,
so we should be able to use that for Karma.

### CODE REVIEW

File: tools/broccoli/gulp/index.ts [hunk 1]
```diff
@@ -26,31 +26,23 @@ function broccoliBuild(tree, outputRoot) {
...
-          time('Write build output', function() {
-            copyDereferenceSync(path.join(dir, outputRoot), distPath);
-          });
-        } catch (err) {
-          if (err.code === 'EEXIST') err.message += ' (we cannot build into an existing directory)';
-          throw err;
-        }
-      })
-      .finally(function() {
-        time('Build cleanup', function() { builder.cleanup(); });
```
Comment (IgorMinar) [inline tools/broccoli/gulp/index.ts]: I believe that we do need to call the cleanup somewhere when exiting

File: tools/broccoli/traceur/index.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ class TraceurFilter extends Writer {
                 // TODO: we should fix the sourceMappingURL written by Traceur instead of overriding
                 // (but we might switch to typescript first)
                 var mapFilepath = filepath.replace(/\.\w+$/, '') + this.destSourceMapExtension;
-                result.js = result.js + '\n  //# sourceMappingURL=./' + path.basename(mapFilepath);
+                result.js = result.js + '\n//# sourceMappingURL=./' + path.basename(mapFilepath);
```
Comment (IgorMinar) [inline tools/broccoli/traceur/index.ts]: accidental change?

File: tools/broccoli/broccoli-dest-copy.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import path = require('path');
+
+/**
+ * Intercepts each file as it is copied to the destination tempdir,
+ * and tees a copy to the given path outside the tmp dir.
+ */
+class DestCopy extends Filter {
+  constructor(private inputTree, private outputRoot: string) { super(inputTree); }
+
+  getDestFilePath(relativePath: string): string { return relativePath; }
```
Comment (IgorMinar) [inline tools/broccoli/broccoli-dest-copy.ts]: do you need to implement these even if they don't do anything?

same for `processString`

**Comment by IgorMinar** [comment]
the rest looks good

File: tools/broccoli/gulp/index.ts (hunk 1 above)
Comment (alexeagle) [inline tools/broccoli/gulp/index.ts]: yup, it's there on line 49 (clang-format decided to shorten to single line)

File: tools/broccoli/traceur/index.ts (hunk 2 above)
Comment (alexeagle) [inline tools/broccoli/traceur/index.ts]: reverting an earlier accidental change :)

File: tools/broccoli/broccoli-dest-copy.ts (hunk 3 above)
Comment (alexeagle) [inline tools/broccoli/broccoli-dest-copy.ts]: yes.
for `destFilePath`: we need to override the superclass behavior of matching on particular extensions
for `processString`: this method is 'abstract' in the superclass
for `processFile`: this is where we need to hook into the spot that knows the absolute paths inside the tmpdir

**Comment by alexeagle** [comment]
merged as 70433e6b73150d29d20af04e0e8c9188459836f6

**Pull request closed without merging**